modules/adnuntiusBidAdapter.js

```diff
@@ -154,28 +154,66 @@ const storageTool = (function () {
   };
 
-  const getUsi = function (meta, ortb2, bidParams) {
-    // Fetch user id from parameters.
-    for (let i = 0; i < bidParams.length; i++) {
-      const bidParam = bidParams[i];
-      if (bidParam.userId) {
-        return bidParam.userId;
-      }
+  const getFirstValidValueFromArray = function(arr, param) {
+    const example = (arr || []).find((b) => {
+      return deepAccess(b, param);
+    });
+    return example ? deepAccess(example, param) : undefined;
+  };
+
+  const normaliseLegacyIds = function(obj) {
+    const result = {};
+    if (!isPlainObjectWithKeys(obj)) {
+      return result;
     }
-    if (ortb2 && ortb2.user && ortb2.user.id) {
-      return ortb2.user.id
+    for (const [key, value] of Object.entries(obj)) {
+      if (isPlainObjectWithKeys(value)) {
+        const theId = value[Object.keys(value)[0]];
+        if (isStr(theId)) {
+          result[key] = theId;
+        }
+      } else if (isStr(value)) {
+        result[key] = value;
+      }
     }
-    return (meta && meta.usi) ? meta.usi : false
-  }
+
+    return result;
+  };
 
   return {
-    refreshStorage: function (bidderRequest) {
-      const ortb2 = bidderRequest.ortb2 || {};
+    refreshStorage: function (validBidRequests, bidderRequest) {
       const bidParams = (bidderRequest.bids || []).map((b) => {
         return b.params ? b.params : {};
       });
       metaInternal = getMetaDataFromLocalStorage(bidParams).reduce((a, entry) => ({ ...a, [entry.key]: entry.value }), {});
-      metaInternal.usi = getUsi(metaInternal, ortb2, bidParams);
+      const bidParamUserId = getFirstValidValueFromArray(bidParams, 'userId');
+      const ortb2 = bidderRequest.ortb2 || {};
+
+      if (isStr(bidParamUserId)) {
+        metaInternal.usi = bidParamUserId;
+      } else if (isStr(ortb2?.user?.id)) {
+        metaInternal.usi = ortb2.user.id;
+      } else {
+        const unvettedOrtb2Eids = deepAccess(ortb2, 'user.ext.eids');
+        const vettedOrtb2Eids = isArray(unvettedOrtb2Eids) && unvettedOrtb2Eids.length > 0 ? unvettedOrtb2Eids : false;
+        const unvettedBidRequestEids = getFirstValidValueFromArray(validBidRequests, 'userIdAsEids') || bidderRequest?.userIdAsEids;
+        const vettedBidRequestEids = isArray(unvettedBidRequestEids) && unvettedBidRequestEids.length > 0 ? unvettedBidRequestEids : false;
+        const bidRequestUserId = getFirstValidValueFromArray(validBidRequests, 'userId') || bidderRequest?.userId;
+
+        if (vettedOrtb2Eids) {
+          metaInternal.eids = vettedOrtb2Eids;
+        } else if (vettedBidRequestEids) {
+            metaInternal.eids = vettedBidRequestEids;
+        } else if (isPlainObjectWithKeys(bidRequestUserId)) {
+          const normalisedLegacyIds = normaliseLegacyIds(bidRequestUserId);
+          if (isPlainObjectWithKeys(normalisedLegacyIds)) {
+            metaInternal.eids = _map(normalisedLegacyIds, function(v, k) {
+              return {source: k, uids: [{id: v}]}
```

Comment:
as @patmmccann says, a lot of this is unncessary - you can just pick `ortb2.user.ext.eids`. This however is also incorrect, `userId` keys are not the EID source but arbitrary labels only used within pbjs.

Reply:
My understanding was that legacy formats of userId can be like this:

```
bidRequest.userId = {
  tdid: "TDID-123",
  pubcid: "PUBCID-456",
  id5id: { uid: "ID5-789" }
};
```

That's why the code was there to convert that into `eids`. And as far as I can tell, they looked like reasonable `source` values even if they are not the standard ones. `pubcid.org` is a standard ortb2 source value, but `pubcid` seems like a decent variant that we can get our adserver to treat the same way for any legacy implementations. 

But if none of that is necessary, that's fine -- happy to simplify and just use `ortb2.user.ext.eids`.